promql/engine.go

```diff
@@ -2141,3 +2302,128 @@ func unwrapParenExpr(e *parser.Expr) {
 		}
 	}
 }
+
+// WrapWithStepInvariantExpr wraps all possible parts of the given
+// expression with StepInvariantExpr wherever valid.
+func WrapWithStepInvariantExpr(expr parser.Expr) parser.Expr {
+	isStepInvariant := wrapWithStepInvariantExprHelper(expr)
+	if isStepInvariant {
+		return newStepInvariantExpr(expr)
+	}
+	return expr
+}
+
+// wrapWithStepInvariantExprHelper wraps the child nodes of the expression
+// with a StepInvariantExpr wherever valid. The returned boolean is true if the
+// passed expression qualifies to be wrapped by StepInvariantExpr.
+func wrapWithStepInvariantExprHelper(expr parser.Expr) bool {
```

Comment:
Do you really need all this logic? Surely you only need to catch the top level node and then use a new evaluator from there?

As is stands it looks like you'll create a new evaluator on every single node within an @ which is wasteful - and possibly on things outside it where you shouldn't.

Reply:
> As is stands it looks like you'll create a new evaluator on every single node within an @ which is wasteful - and possibly on things outside it where you shouldn't.

It will create new evaluator at the top most possible level where it's a step invariant. For example if I had `rate(foo[1m] @ 10) + rate(bar[1m] @ 10)`, then the entire thing is a single step invariant node since the information is propagated upwards.

So this logic is useful in avoid as much repeated calculations as possible.

Reply:
I will be adding tests for this wrapping stuff

Reply:
That's over-optimising, and in a way that will be practically impossible to be correct.

> It will create new evaluator at the top most possible level where it's a step invariant. 

That's not what the code says.

Keep it simple, eliminate all of this code.

Reply:
> Keep it simple, eliminate all of this code.

Then the use-case for which this syntax is being added is going to be very inefficient. Re-calculating the same result for every step does not sound ideal.

Reply:
Get the code correct first, then worry about optimisations. Once we've a good set of unittests it'll be much easier to consider if this sort of thing is actually workable.